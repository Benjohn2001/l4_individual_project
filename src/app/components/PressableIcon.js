import { Component } from "react";
import PropTypes from 'prop-types';
import { Pressable, TouchableOpacity, View } from "react-native";
import { Feather } from '@expo/vector-icons';

class PressableIcon extends Component {
    render(){
        const { onPress , icon, size, color } = this.props;
		return (
            <View className="flex-row">
                <Pressable onPress={onPress}>
                        <Feather name={icon} size={size} color={color} />
                </Pressable>
            </View>
        );
    }
}

PressableIcon.propTypes = {
    onPress: PropTypes.func.isRequired,
    icon: PropTypes.string.isRequired,
    size: PropTypes.number.isRequired,
    color: PropTypes.string.isRequired
};
        
export default PressableIcon;