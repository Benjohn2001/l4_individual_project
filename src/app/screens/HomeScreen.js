import React from 'react';
import { Feather } from '@expo/vector-icons';
import { View, Image, TextInput, Text, TouchableOpacity, SafeAreaView, ScrollView, FlatList, ActivityIndicator} from 'react-native';
import AppButtonPurple from '../components/AppButtonPurple'
import AppButtonLight from '../components/AppButtonLight';
import TwoButtonsSide from '../components/TwoButtonsSide';
import { COLOURS } from '../assets/colours';
import GroupButton from '../components/GroupButton';
import PressableIcon from '../components/PressableIcon';
import Modal from "react-native-modal";
import TwoButtonStack from '../components/TwoButtonStack';
import { auth } from '../../firebase';
import { query, get, ref, getDatabase } from 'firebase/database';
import SearchBar from "react-native-dynamic-search-bar";


const HomeScreen = ({ navigation }) => {
    
    const [isModalVisible, setModalVisible] = React.useState(false)
    const [groups, setGroups] = React.useState([])
    const [fetched, setFetched] = React.useState(false)
    const [searchVis, setSearchVis] = React.useState(false)
    const [spinVis, setSpinVis] = React.useState(false)
    const [filtered, setFiltered] = React.useState([])

    React.useEffect(() =>{
          navigation.addListener('beforeRemove', (e) => {
            e.preventDefault();
        })
        fetchData()
    },[] )

    const fetchData = async () =>{
        setGroups([])
        const groupsRef = query(ref(getDatabase(), "/groups/"+auth.currentUser.uid))
        const data = await get(groupsRef)
        data.forEach(c => {
            setGroups(a => {return [...a , c]})
        })
        setFetched(true)
    }

    const filterGroups = (name) => {
        setFiltered([])
        groups.filter((str)=>{
                if(str.val()["name"].toLowerCase().includes(name.toLowerCase())){
                    setFiltered(a => {return [...a , str]}) 
                }
        })
    }

    return (
        <SafeAreaView className="flex-1 bg-primaryPurple" >
            <View className="flex-row justify-between pt-14 mx-10">
                <Text className="font-bold text-3xl">
                    My Groups
                </Text>
                <PressableIcon
                    onPress={() => {
                        searchVis ? setSearchVis(false) : setSearchVis(true)
                    }}
                    icon="search"
                    size={40}
                    color="black"
                />
            </View>
            <View className="flex-1 justify-center items-center pt-10">
            {
                searchVis ?
                <View>
                    <SearchBar
                        className="bg-secondaryPurple  h-12 w-11/12"
                        placeholder="Search for a group"
                        iconColor="6B4EFF"
                        spinnerVisibility={spinVis}
                        onChangeText={text => {
                            setFiltered([])
                            if (text.length === 0) {
                                setSpinVis(false);
                            } else {
                                setSpinVis(true);
                            }
                            filterGroups(text);
                        } }

                        onClearPress={() => {
                            filterGroups('');
                            setSpinVis(false);
                        } } />
                        <View>
                            <FlatList
                                data={filtered}
                                renderItem={ ({item})=>(
                                    <GroupButton
                                        groupName={item.val()["name"]}
                                        onPress={() => {
                                            navigation.navigate("GroupScreen",{
                                                name: item.val()["name"]
                                            })
                                        }}
                                        avatar={require('../assets/ben-avatar.png')}
                                    />
                                )

                            
                                }
                            />
                        </View>
                </View>
                :
            <View>
                {fetched ?
                    <FlatList
                            showsVerticalScrollIndicator={false}
                            extraData={groups}
                            data={groups}
                            renderItem={({ item }) => (
                                <GroupButton
                                groupName={item.val()["name"]}
                                onPress={() => {
                                    navigation.navigate("GroupScreen",{
                                        name: item.val()["name"]
                                    })
                                }}
                                avatar={require('../assets/ben-avatar.png')}
                            />
                            )} 
                        />
                    :
                    <View className="justify-center items-center flex-1">
                        <ActivityIndicator size="large" color="#6B4EFF"  />
                        <Text className="text-center">Loading Groups</Text>
                    </View>
                }
            </View>
            }       
            </View>
            
            <View className="items-center py-3">
                <AppButtonPurple
                    title="Add Another Group"
                    onPress={() => {
                        setModalVisible(true)
                    }
                    }
                />
            </View>
            <Modal 
                isVisible={isModalVisible}
                onBackdropPress={() => setModalVisible(false)}
                className="items-center"
            >
                <View className=" w-11/12 h-1/2 items-center py-3 px-3 bg-secondaryPurple rounded-2xl">
                    <Text className="text-2xl font-bold  pb-2">Add a group</Text>
                    <Text className="text-center text-gray-500 py-5">Add with invite code or create a new group</Text>
                    <TextInput 
                        className="bg-white my-5 w-full mx-3 h-12 rounded-md" 
                        placeholder="Invite Code" 
                        underlineColorAndroid = "transparent"
                        cursorColor={COLOURS.darkerPurple}
                    />
                    <View className="pt-5">
                        <TwoButtonStack
                            title1="Add Group"
                            onPress1={() => {
                                alert("Add Group")
                            }}
                            title2="Create New Group"
                            onPress2={() => {
                                navigation.navigate("CreateNewGroupScreen")
                            }}
                        />
                    </View>
                </View>
            </Modal>
         
        </SafeAreaView>

    );
};

export default HomeScreen