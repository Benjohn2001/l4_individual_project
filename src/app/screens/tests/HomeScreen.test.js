import renderer from "react-test-renderer";
import { render, screen, fireEvent } from "@testing-library/react-native";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import HomeScreen from "../HomeScreen";
import PressableIcon from "../../components/PressableIcon";

configure({ adapter: new Adapter() });

jest.mock("firebase/database", () => ({
  getDatabase: jest.fn().mockReturnThis(),
  push: jest.fn().mockReturnThis(),
  ref: jest.fn().mockReturnThis(),
  set: jest.fn().mockReturnThis(),
}));

jest.mock("firebase/storage", () => ({
  getStorage: jest.fn().mockReturnThis(),
}));

const navigation = {
  navigate: jest.fn(),
  push: jest.fn(),
  addListener: jest.fn().mockImplementation((event, callback) => {
    if (event === "beforeRemove") {
      callback({ preventDefault: jest.fn() });
    }
  }),
};

jest.mock("firebase/app", () => ({
  auth: jest.fn().mockReturnThis(),
  initializeApp: jest.fn().mockReturnThis(),
}));

jest.mock("@react-navigation/native", () => ({
  useIsFocused: jest.fn(),
}));

jest.mock("firebase/auth", () => ({
  auth: jest.fn().mockReturnThis(),
  getAuth: () => ({
    signOut: () => ({
      then: () => ({
        catch: jest.fn(),
      }),
    }),
  }),
}));

test("Check HomeScreen search page correctly", () => {
  const wrapper = shallow(<HomeScreen navigation={navigation} />);
  const icon = wrapper.find(PressableIcon);
  icon.simulate("press");
  expect(wrapper).toMatchSnapshot();
});

test("Check HomeScreen loading correctly", () => {
  render(<HomeScreen navigation={navigation} />);
  const domTree = renderer.create(<HomeScreen />).toJSON();
  expect(domTree).toMatchSnapshot();
});

test("Check HomeScreen search correctly", () => {
  render(<HomeScreen navigation={navigation} />);
  fireEvent.press(screen.getByTestId("pressableIcon"));
  const searchInput = screen.getByPlaceholderText("Search for a group");

  fireEvent.changeText(searchInput, "group");
  fireEvent.changeText(searchInput, "");
});

test("Check HomeScreen add group correctly", () => {
  render(<HomeScreen navigation={navigation} />);
  fireEvent.press(screen.getByText("Add a Group"));
  fireEvent.press(screen.getByText("Create New Group"));
  expect(navigation.push).toBeCalledWith("CreateNewGroupScreen");
});
