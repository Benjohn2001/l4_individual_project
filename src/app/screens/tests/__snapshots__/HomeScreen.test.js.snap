// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check HomeScreen loading correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      Object {
        "backgroundColor": "#C6C4FF",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
        },
        Object {
          "justifyContent": "space-between",
        },
        Object {
          "paddingTop": 56,
        },
        Object {
          "marginLeft": 40,
          "marginRight": 40,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontWeight": "700",
          },
          Object {
            "fontSize": 30,
            "lineHeight": 36,
          },
        ]
      }
    >
      My Groups
    </Text>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
        ]
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="pressableIcon"
      >
        <Text />
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        Object {
          "justifyContent": "center",
        },
        Object {
          "alignItems": "center",
        },
        Object {
          "paddingTop": 40,
        },
      ]
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "justifyContent": "center",
            },
            Object {
              "alignItems": "center",
            },
            Object {
              "flexBasis": "0%",
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      >
        <ActivityIndicator
          color="#6B4EFF"
          size="large"
        />
        <Text
          style={
            Array [
              Object {
                "textAlign": "center",
              },
            ]
          }
        >
          Loading Groups
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
        Object {
          "paddingBottom": 12,
          "paddingTop": 12,
        },
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#6B4EFF",
          "borderBottomLeftRadius": 9999,
          "borderBottomRightRadius": 9999,
          "borderTopLeftRadius": 9999,
          "borderTopRightRadius": 9999,
          "height": 48,
          "marginLeft": 6,
          "marginRight": 6,
          "opacity": 1,
          "paddingTop": 12,
          "width": "83.333333%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#fff",
            },
            Object {
              "textAlign": "center",
            },
          ]
        }
      >
        Add a Group
      </Text>
    </View>
  </View>
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    transparent={true}
    visible={false}
  />
</RCTSafeAreaView>
`;

exports[`Check HomeScreen search page correctly 1`] = `ShallowWrapper {}`;
